### PR TITLE
NDRS-970: Remove semver crate and use ProtocolVersion

### PR DESCRIPTION
**Pull request by darthsiroftardis**
REF: https://casperlabs.atlassian.net/browse/NDRS-970

CHANGELOG:

- Removed `semver` crate from `node`, `client` and `ci` dependencies
- Replaced usages of the crate with the `casper-types::ProtocolVersion` instead.
- Added a new `FromStr` impl for `ProtocolVersion` to help with refactoring
- Refactored code to use `ProtocolVersion`

### CODE REVIEW

**Review by Fraser999 — COMMENTED** [review]
Only some nitpicks - looking good otherwise.

**Comment by goral09** [comment]
Can we, as part of this PR, add `impl From<Version> for ProtocolVersion` ?

**Comment by Fraser999** [comment]
> Can we, as part of this PR, add impl From<Version> for ProtocolVersion ?

That won't be required now as this PR removes all instances of `semver::Version` from our crates.

File: node/src/components/rest_server/filters.rs [hunk 1]
```diff
@@ -26,18 +27,18 @@ pub const METRICS_API_PATH: &str = "metrics";
...
 pub(super) fn create_status_filter<REv: ReactorEventT>(
     effect_builder: EffectBuilder<REv>,
-    api_version: Version,
+    api_version: ProtocolVersion,
 ) -> BoxedFilter<(Response<Body>,)> {
     warp::get()
         .and(warp::path(STATUS_API_PATH))
         .and_then(move || {
-            let api_version_cloned = api_version.clone();
+            let api_version_cloned = api_version;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rest_server/filters.rs:35]: Yeah, the linter had caught it in a bunch of places, but I guess it missed a few

File: ci/casper_updater/src/package.rs [hunk 2]
```diff
@@ -1,10 +1,10 @@
+use core::str::FromStr;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply ci/casper_updater/src/package.rs]: Using `std`

File: ci/casper_updater/src/package.rs [hunk 3]
```diff
@@ -15,6 +15,8 @@ use crate::{
     BumpVersion,
 };
 
+use casper_types::ProtocolVersion;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply ci/casper_updater/src/package.rs]: Using `SemVer`

File: node/src/components/event_stream_server/sse_server.rs [hunk 4]
```diff
@@ -27,11 +26,12 @@ type Id = u32;
 
 /// The "data" field of the events sent on the event stream to clients.
 #[derive(Clone, PartialEq, Eq, Serialize, Deserialize, Debug, DataSize)]
+#[allow(clippy::large_enum_variant)]
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/event_stream_server/sse_server.rs]: Reduced variant size difference

File: node/src/components/rest_server/filters.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rest_server/filters.rs]: Using simply `api_version`

File: node/src/components/rpc_server/rpcs.rs [hunk 5]
```diff
@@ -85,15 +86,15 @@ pub(super) trait RpcWithParamsExt: RpcWithParams {
...
+        api_version: ProtocolVersion,
     ) -> BoxedFilter<(Response<Body>,)> {
         let with_valid_params = warp::path(RPC_API_PATH)
             .and(filters::json_rpc())
             .and(filters::method(Self::METHOD))
             .and(filters::params::<Self::RequestParams>())
             .and_then(
                 move |response_builder: Builder, params: Self::RequestParams| {
-                    let api_version_cloned = api_version.clone();
+                    let api_version_cloned = api_version;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs.rs]: Same as above

File: node/src/components/rpc_server/rpcs.rs [hunk 6]
```diff
@@ -217,9 +218,9 @@ pub(super) trait RpcWithOptionalParamsExt: RpcWithOptionalParams {
     /// Creates the warp filter for this particular RPC.
     fn create_filter<REv: ReactorEventT>(
         effect_builder: EffectBuilder<REv>,
-        api_version: Version,
+        api_version: ProtocolVersion,
     ) -> BoxedFilter<(Response<Body>,)> {
-        let api_version_cloned = api_version.clone();
+        let api_version_cloned = api_version;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs.rs]: And again.

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 7]
```diff
@@ -27,8 +28,8 @@ use super::{
 };
 use crate::{effect::EffectBuilder, rpcs::chain::GetEraInfoBySwitchBlock};
 
-pub(crate) static DOCS_EXAMPLE_PROTOCOL_VERSION: Lazy<Version> =
-    Lazy::new(|| Version::new(1, 0, 0));
+pub(crate) static DOCS_EXAMPLE_PROTOCOL_VERSION: Lazy<ProtocolVersion> =
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs]: using a `const`

File: node/src/components/storage.rs [hunk 8]
```diff
@@ -292,11 +291,7 @@ impl Storage {
 
         // Note: `iter_start` has an undocumented panic if called on an empty database. We rely on
         //       the iterator being at the start when created.
-        let protocol_version = ProtocolVersion::from_parts(
-            version.major as u32,
-            version.minor as u32,
-            version.patch as u32,
-        );
+        let protocol_version = version;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/storage.rs]: Renamed parameter, and deleted line

File: node/src/types/chainspec/protocol_config.rs [hunk 9]
```diff
@@ -1,13 +1,17 @@
 // TODO - remove once schemars stops causing warning.
 #![allow(clippy::field_reassign_with_default)]
 
+use core::str::FromStr;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/chainspec/protocol_config.rs]: Using `std::str`

File: node/src/components/chainspec_loader.rs [hunk 10]
```diff
@@ -372,16 +373,16 @@ impl ChainspecLoader {
 
         let debug_assert_version_match = || {
             debug_assert!(
-                previous_protocol_version.value().major as u64
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/chainspec_loader.rs:375]: using one `debug_assert`

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**